include/fiction/technology/sidb_cluster_hierarchy.hpp

```diff
@@ -0,0 +1,1083 @@
+     * passed separately.
+     *
+     * @param loc_pot_min The minimum local potential for the SiDB in the singleton cluster.
+     * @param loc_pot_max The maximum local potential for the SiDB in the singleton cluster.
+     * @param loc_ext_pot The local external potential at the SiDB in the singleton cluster. Specifically, this is the
+     * sum of the local defect potential and the local external potential.
+     * @param base The simulation base.
+     * @param self_ptr Shared pointer to itself.
+     */
+    void initialize_singleton_cluster_charge_space(const double loc_pot_min, const double loc_pot_max,
+                                                   const double loc_ext_pot, const uint8_t base,
+                                                   const sidb_cluster_ptr& self_ptr) noexcept
+    {
+        assert(sidbs.size() == 1);
+
+        const uint64_t ix = *sidbs.cbegin();
+
+        for (const sidb_charge_state cs : SIDB_CHARGE_STATES(base))
+        {
+            charge_space.emplace(self_ptr, cs, loc_ext_pot);
+        }
+
+        received_ext_pot_bounds.set(ix, loc_pot_min, loc_pot_max);
+    }
+    /**
+     * Function to return the number of SiDBs contained in the cluster.
+     *
+     * @return The number of SiDBs contained in the cluster.
+     */
+    inline uint64_t num_sidbs() const noexcept
+    {
+        return sidbs.size();
+    }
+    /**
+     * Equality operation on cluster hierarchies. Checks the respective unique identifiers for equality.
+     *
+     * @param other Cluster (hierarchy) to compare to.
+     * @return `true` if and only if the unique identifiers match.
+     */
+    constexpr inline bool operator==(const sidb_cluster& other) const noexcept
+    {
+        return uid == other.uid;
+    }
+};
+/**
+ * Function to obtain the amount of SiDBs in the given cluster. Required for compilation due to mutual recursion.
+ *
+ * @param c Cluster of which the number of SiDBs it contains is requested.
+ * @return The number of SiDBs in the given cluster.
+ */
+static inline uint64_t get_cluster_size(const sidb_cluster_ptr& c) noexcept
+{
+    return c->sidbs.size();
+}
+/**
+ * Function to obtain the unique identifier of the given cluster. Required for compilation due to mutual recursion.
+ *
+ * @param c Cluster of which its unique identifier is requested.
+ * @return The unique identifier of the given cluster.
+ */
+static inline uint64_t get_unique_cluster_id(const sidb_cluster_ptr& c) noexcept
+{
+    return c->uid;
+}
+/**
+ * Function to obtain the SiDB index contained in a singleton cluster. Required for compilation due to mutual recursion.
+ *
+ * @param c Singleton cluster of which the single SiDB (index) it contains is requested.
+ * @return The SiDB index contained in the given cluster. It is equal to the unique identifier of the cluster.
+ */
+static inline uint64_t get_singleton_sidb_ix(const sidb_cluster_ptr& c) noexcept
+{
+    assert(get_cluster_size(c) == 1);
```

Comment:
please add a short text.

Reply:
There is a docstring right?

Reply:
Ah, you mean for the `assert`

Reply:
Sorry, I was imprecise: I meant inside the assert. For example:
```c++
assert(get_cluster_size(c) == 1 && "Cluster size is not equal to 1");
```